lib_eio/eio.mli

```diff
@@ -740,18 +749,33 @@ module Net : sig
     sw:Switch.t ->
     #listening_socket ->
     on_error:(exn -> unit) ->
-    (sw:Switch.t -> <Flow.two_way; Flow.close> -> Sockaddr.t -> unit) ->
+    (sw:Switch.t -> <Flow.two_way; Flow.close> -> Sockaddr.stream -> unit) ->
     unit
-  (** [accept socket fn] waits for a new connection to [socket] and then runs [fn ~sw flow client_addr] in a new fibre,
+  (** [accept_sub socket fn] waits for a new connection to [socket] and then runs [fn ~sw flow client_addr] in a new fibre,
       using {!Fibre.fork_on_accept}.
       [flow] will be closed automatically when the sub-switch is finished, if not already closed by then. *)
 
+  class virtual ['addr] datagram_socket : object
+    method virtual send : 'addr -> Cstruct.t -> unit
+    method virtual recv : Cstruct.t -> 'addr * int
+  end
+
+  val send : 'addr datagram_socket -> 'addr -> Cstruct.t -> unit
+  (** [send sock addr buf] sends the data in [buf] to the address [addr] using the 
+      the socket [sock]. *)
+
+  val recv : 'addr datagram_socket -> Cstruct.t -> 'addr * int
+  (** [recv sock buf] receives data from the socket [sock] putting it in [buf]. The number of bytes received is 
+      returned along with the sender address and port. If the [buf] is too small then excess bytes may be discarded
+      depending on the type of the socket the message is received from. *)
+
   class virtual t : object
-    method virtual listen : reuse_addr:bool -> reuse_port:bool -> backlog:int -> sw:Switch.t -> Sockaddr.t -> listening_socket
-    method virtual connect : sw:Switch.t -> Sockaddr.t -> <Flow.two_way; Flow.close>
+    method virtual listen : reuse_addr:bool -> reuse_port:bool -> backlog:int -> sw:Switch.t -> Sockaddr.stream -> listening_socket
+    method virtual connect : sw:Switch.t -> Sockaddr.stream -> <Flow.two_way; Flow.close>
+    method virtual datagram_socket : sw:Switch.t -> Sockaddr.datagram -> Sockaddr.datagram datagram_socket
```

Comment:
When I suggested the type parameter, I was thinking of something like this:
```suggestion
    method virtual datagram_socket : 'a. sw:Switch.t -> ([< Sockaddr.datagram] as 'a) -> 'a datagram_socket
```
If we only want UDP datagrams, the type could be simplified though. According to my Linux's unix(7) man-page you can use Unix datagrams for reliable datagrams that preserve message boundaries, though I don't think I've ever used this.